### DISCUSSION TITLE
actions as data

### DISCUSSION THREAD

**Discussion by jmagaram** [statelyai/xstate#2043] [Q&A]
I'm trying to create a state machine that is "pure" / self-contained.

For guards, it seems like there are a couple options. You can define the guard as a string, like `userIsAdministrator` and then in the second parameter to the Machine function you can provide an implementation for that function, like `(ctx,evt)=>evt.User.IsAdministrator`. One problem with this approach is that the TypeScript isn't set up to provide a narrowed event, presumably because `userIsAdministrator` could be used for many different guards with many different events. So I need to cast the event to a specific event before using it, which isn't type safe. So instead I just put arrow functions throughout the machine and really see no harm in that, since these are just pure functions that rely on the current context and event.

For actions / side effects that modify the context, I do the same thing - sprinkle `assign` statements throughout the machine.

For actions / side effects that touch the outside world, I don't see a good solution. I can hardwire my effects like `(ctx,evt)=>console.log(evt.errorMessage)`, but now my machine is not "pure" / self-contained. If I create a side effect as a string like `displayErrorMessage` then when I define the actual effect in the second parameter to the Machine function I need to again cast the event to one that has an `errorMessage` property and this isn't type safe. I'd prefer to be able to define my side effects in the machine using data, like `{ action: 'displayErrorMessage', message: ctx.ErrorMessage}`. This way the side effect can make full use of the current typed-event . I can imagine the Machine having a new type parameter called `Action`, where every action must conform to that type specification. And then in the interpreter, or maybe in the second parameter to the Machine function, there would be some way to map these "actions as data" to executable statements. So when I see an action of type `displayErrorMessage` I could do the appropriate console.log or whatever.

I'm just getting started with xState but am wondering whether this notion of creating a "pure" machine is desirable. And if so, is there a way to encode actions as data? Is there a way to prevent having to do the unsafe casting of events in the second parameter to the Machine function? I see something called a `pure action` - maybe this solves my problem but I couldn't figure it out.

**Comment by Andarist** [comment]
> One problem with this approach is that the TypeScript isn't set up to provide a narrowed event, presumably because userIsAdministrator could be used for many different guards with many different events.

Yep, that's the problem - we can't easily determine where the string has been used to provide narrowed down event type for the implementation of a certain thing.

> So instead I just put arrow functions throughout the machine and really see no harm in that, since these are just pure functions that rely on the current context and event.

Yes, no real harm in that - apart from the fact that your schema is not fully serializable right now but that's also not always needed.

> For actions / side effects that touch the outside world, I don't see a good solution. I can hardwire my effects like (ctx,evt)=>console.log(evt.errorMessage), but now my machine is not "pure" / self-contained.

I don't understand the problem here. It's the very same situation as you have with guards, assign actions, or whatever. Those functions are not invoked right away. It doesn't really matter for the runtime if things are defined inline or referenced using strings.

> Is there a way to prevent having to do the unsafe casting of events in the second parameter to the Machine function?

There is an experimental [`xstate-codegen`](https://github.com/mattpocock/xstate-codegen) on which we are going to expand in the future and incorporate that into the official tooling.

>  I see something called a pure action - maybe this solves my problem but I couldn't figure it out.

This is about something different - it allows you to create actions dynamically, using the full power of JS (rather than having to use constrained action creators like `choose`)

**Reply by jmagaram** [reply]
Typescript is pretty powerful. I wouldn't be surprised if there was a way to use TypeScript to find all the places where a particular guard/effect string is used and do a union on all the events to make it type safe.

I guess I'm not sure what code to put inside my machine for the effects while I'm designing it. I can put in the real code to do things like display dialog boxes etc. But I'd prefer to wait to define these later and to decouple the real implementation. Putting in a string like 'displayErrorMessage' isn't good enough since it doesn't include the data I need from the actual event. I could convert all my actions to something like `()=>execute({action:'displayErrorMessage'; message:ctx.error})` and create a stub/empty `execute` function while prototyping, and insert the real `execute` function later. Hmmm...

**Reply by jmagaram** [reply]
I'm hoping to achieve (a) a machine that works well in the visualizer - can see short descriptions of guards and effects, (b) a machine I can "prototype" without coding real effects - quickly typing actions as strings like `showErrorMessage` is useful, (c) type safety, (d) a machine where the definition is easy to read and understand - the less code the better. It doesn't look like I can achieve all of these things.

If I use strings for the guards and effects, then +abd but -c.

If I type real descriptive function names for the guards and effects, then +cd but -ab and assigning the Typescript types to these standalone functions is painful.

If I type arrow functions for guards and effects within the machine, then +c but -abd.

If there was some way to type the guards and effects as data, and later customize the interpreter to execute these effects and evaluate the guards, I think I could have all +abcd. For example, an effect could be encoded as an object like `{ action:'showMessage'; message: evt=>evt.error }`. Guards could be encoded with custom objects like `{ cond:'areEmailsTheSame', a: ctx=>ctx.email, b: evt=>evt.email }`. The visualizer could display the discriminator text. Or these custom objects could have a "comment" property for the visualizer.

A state machine compiler or better Typescript or a DSL would help.

**Reply by davidkpiano** [reply]
This is in the works with the `machine.schema` field, by the way. Soon, you should be able to provide types for actions:

```ts
const machine = createMachine({
  // ...
  schema: {
    actions: {
      showMessage: createSchema<{ message: string }>(),
      // ...
    }
  }
}
```

Right now, actions _aren't_ inferred throughout the machine, and `schema.actions` doesn't help strongly type the actions (yet), but it's definitely possible.

**Reply by Andarist** [reply]
> Typescript is pretty powerful. I wouldn't be surprised if there was a way to use TypeScript to find all the places where a particular guard/effect string is used and do a union on all the events to make it type safe.

It is somewhat possible - but the TS code doing it is *crazy*. [This project](https://github.com/devanshj/txstate) is pushing TS and XState API to its limits but it's incomplete yet (it doesn't yet solve this problem that you mention but it's solving harder problems already so my bet is that this is possible as well). However, apart from the correctness we also need to care about maintainability and performance and without doing a lot of testing (and finishing the project) we can't really say that this will be possible to do using just TS.

>  But I'd prefer to wait to define these later and to decouple the real implementation.

Sure, this makes sense - I just don't get why you are OK with some things (guards, assign actions) being defined inline but not for everything. It seems that you'd like to define all of the things "externally".

> If I type arrow functions for guards and effects within the machine, then +c but -abd.

That's not completely true because this only works for transition actions (and not for entry/exit actions and it also doesn't work correctly for eventless transitions). So this only partially delivers what you want

> A state machine compiler or better Typescript or a DSL would help.

This is somewhat a long-term plan and the mentioned xstate-codegen is the first step towards that.

> This is in the works with the machine.schema field, by the way. Soon, you should be able to provide types for actions:

This issue here is not as much about having action strings type-checked as it is about correctly narrowing down `TEvent` to those event types that can actually be received by any particular guard/action/etc. So schema thing doesn't actually address it - nor it has the potential of doing it.

**Reply by devanshj** [reply]
> so my bet is that this is possible as well

You're making good bets @Andarist :P

And btw +1 on this -
> So schema thing doesn't actually address it - nor it has the potential of doing it.
